src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1728,17 +1728,19 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
                 keyword = ConvertToKeyword(this.EatToken());
             }
 
+            bool isExtension = keyword.Kind == SyntaxKind.ExtensionKeyword;
             var outerSaveTerm = _termState;
-            _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceSignature;
+            _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceOrExtensionSignature;
 
             var saveTerm = _termState;
             _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop;
 
-            var name = this.ParseIdentifierToken();
+            SyntaxToken? name = (isExtension && this.CurrentToken.Kind != SyntaxKind.IdentifierToken) ? null : this.ParseIdentifierToken();
+
             var typeParameters = this.ParseTypeParameterList();
 
-            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken
-                ? ParseParenthesizedParameterList() : null;
+            var paramList = CurrentToken.Kind == SyntaxKind.OpenParenToken || isExtension
+                ? ParseParenthesizedParameterList(allowOptionalIdentifier: isExtension) : null;
```

Comment:
nit: doc this.  probably something like: In the case of extensions, there must always be a parameter list.  Requiring this here also means that later parts of the compiler do not have to be resilient to getting back a null parameter list from teh declaration.